internal/sql/pg/system.go

```diff
@@ -198,6 +202,452 @@ var settingValidations = map[string]settingValidFn{
+	baseType := ci.baseType()
+	if ci.Array {
+		return arrayType(baseType)
+	}
+	return baseType
+}
+
+func (ci *ColInfo) baseType() ColType {
+	// TODO: merge into since switch or map when this has settled.
+	if ci.IsInt() {
+		return ColTypeInt
+	}
+	if ci.IsText() {
+		return ColTypeText
+	}
+	if ci.IsBool() {
+		return ColTypeBool
+	}
+	if ci.IsByteA() {
+		return ColTypeByteA
+	}
+	if ci.IsNumeric() {
+		return ColTypeNumeric
+	}
+	if ci.IsUINT256() {
+		return ColTypeUINT256
+	}
+	if ci.IsFloat() {
+		return ColTypeFloat
+	}
+	if ci.IsUUID() {
+		return ColTypeUUID
+	}
+	if ci.IsTime() {
+		return ColTypeTime
+	}
+	return ColTypeUnknown
+}
+
+// The following methods recognize the DataType values as reported by "regtype"
+// values in the information_schema.columns PostgreSQL system table. Use the
+// Type method to obtain the canonical ColType.
+
+func (ci *ColInfo) IsInt() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "bigint", "integer", "smallint", "int", "int2", "int4", "int8":
+		return true
+	}
+	return false
+}
+
+func (ci *ColInfo) IsText() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "text", "varchar":
+		return true
+	}
+	return false
+}
+
+func (ci *ColInfo) IsFloat() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "double precision", "single precision", "float32", "float64":
+		return true
+	}
+	return false
+}
+
+func (ci *ColInfo) IsBool() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "boolean", "bool":
+		return true
+	}
+	return false
+}
+
+func (ci *ColInfo) IsUUID() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "uuid":
+		return true
+	}
+	return false
+
+}
+
+func (ci *ColInfo) IsByteA() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "bytea":
+		return true
+	}
+	return false
+}
+
+func (ci *ColInfo) IsUINT256() bool {
+	switch strings.ToLower(ci.DataType) {
+	case "uint256":
+		return true
+	}
+	return false
+}
+
+func (ci *ColInfo) IsNumeric() bool {
+	dt := strings.ToLower(ci.DataType)
+	return strings.HasPrefix(dt, "numeric") // all numeric, including plain or with prec/scale
+}
+
+func (ci *ColInfo) IsTime() bool {
+	dt := strings.ToLower(ci.DataType)
+	return strings.HasPrefix(dt, "timestamp") // includes timestamptz
+}
+
+func columnInfo(ctx context.Context, conn *pgx.Conn, schema, tbl string) ([]ColInfo, error) {
+	var colInfo []ColInfo
+
+	if schema == "" {
+		schema = "public" // otherwise we can get multiple rows
+	}
+
+	dbName := conn.Config().Database
+
+	// get column data types
+	sql := `SELECT ordinal_position, column_name,
+			data_type, udt_name::regtype, domain_name::regtype,
+			is_nullable, column_default
+        FROM information_schema.columns
+        WHERE table_name = '` + tbl + `' AND table_schema = '` + schema + `'
+			AND table_catalog = '` + dbName + `'`
+
+	var worked bool
+
+	var pos int
+	var domainName pgtype.Text // null in Valid bool
+	var colName, dataType, typeOrArray, isNullable string
+	var colDefault any
```

Comment:
This default is scanned out, but it will probably be pretty confusing to a consumer of this information. In the `ColInfo` struct, it is of type `any`, however a consumer would probably think that it would be one of Kwil's types (e.g. our numeric type). However, a numeric will actually be one of pgx's.

It seems like this (or the `queryRowFunc`) is a place that we could rely on the `*datatype` struct to convert this, using the `(*datatype.Decode())` method.

Alternatively, making the `*pg.ColInfo` struct unexported could make sense. Really, what I am getting at here is that the consumer of this package is left with ambiguity as to whether things are converted to be "Kwil native", or if they are "pgx native"

Reply:
I agree that exporting this `colDefault` is unhelpful at best and problematic for the reasons you mentioned.  I didn't give this field much thought, really it was just for completeness but it's not helpful outside this package without more constrained types.

I'll almost certainly unexport things rather than cast `ColInfo` and it's fields in terms of Kwil types.  The `*datatype` and it's `Decode` method are designed to address a different problem -- interpreting the `any` from the `row.Values()` return in `pgx.CollectRows`.  In this work we want to dictate the scan value's type so that it's predicatable and easy to work with, no fuss with OIDs and ambiguity about what type pgx might choose if left to its own devices (or postgres' in the case of domains).  The types from the ScanVal method are quite tightly coupled to the (unexported) `colStats`.  ` TableStats` is the only thing that really uses these.